src/v/pandaproxy/rest/handlers.cc

```diff
@@ -140,6 +140,8 @@ get_topics_records(server::request_t rq, server::reply_t rp) {
       timeout,
       max_bytes);
 
+    auto units
+      = co_await storage::internal::maybe_reserve_decompression_units();
```

Comment:
It is possible to move this nearer to the serialization, I tried in the latest version. It required a few changes but I think it is better because we don't need to hold units while fetching partition. Unfortunately we cannot move it any further as all code is sync once we start json serialization.

I will also tag the enterprise team to have a look if there is a better approach.